### PR TITLE
feat: integrate confidence sequences in automl and epsilon_decay

### PR DESCRIPTION
**Pull request by bassmang**

### CODE REVIEW

File: vowpalwabbit/core/src/metric_sink.cc [hunk 1]
```diff
@@ -73,10 +80,19 @@ bool VW::metric_sink::get_bool(const std::string& key) const
   return it->second;
 }
 
+VW::metric_sink VW::metric_sink::get_metric_sink(const std::string& key) const
+{
+  auto it = _metric_sink_metrics.find(key);
+  if (it == _metric_sink_metrics.end())
+  { THROW("Key: " << key << " does not exist in bool metrics. Is the type correct?") }
```
Comment (jackgerrits) [inline vowpalwabbit/core/src/metric_sink.cc:87]: Is this the right message?

**Review by jackgerrits — APPROVED** [review]

**Comment by lalo** [comment]
don't merge yet

File: vowpalwabbit/core/src/reductions/details/automl/automl_impl.cc [hunk 2]
```diff
@@ -396,20 +391,20 @@ void automl<CMType>::offset_learn(multi_learner& base, multi_ex& ec, CB::cb_clas
   // ** Note: champ learning is done after to ensure correct feature count in gd **
   // Learn and get action of champ
   cm->do_learning(base, ec, current_champ);
-  auto champ_action = ec[0]->pred.a_s[0].action;
+
   for (live_slot = 1; static_cast<size_t>(live_slot) < cm->estimators.size(); ++live_slot)
   {
-    if (cm->_lb_trick) { cm->estimators[live_slot].second.update(champ_action == labelled_action ? w : 0, 1 - r); }
+    if (cm->_lb_trick) { cm->estimators[live_slot].second.update(1, 1 - r); }
```
Comment (lalo) [inline vowpalwabbit/core/src/reductions/details/automl/automl_impl.cc:397]: technically ignores exploration, if used with cb_explore_adf?

File: vowpalwabbit/core/src/reductions/details/automl/automl_oracle.cc [hunk 3]
```diff
@@ -38,6 +38,15 @@ void config_oracle<oracle_impl>::insert_starting_configuration()
   configs.emplace_back(set_ns_list_t(), global_lease, config_type::Exclusion);
   ++valid_config_size;
 }
+template <>
+void config_oracle<champdupe_impl>::keep_best_two(uint64_t winner_config_index)
+{
+  std::swap(configs[0], configs[winner_config_index]);
+  if (winner_config_index != 1) { std::swap(configs[1], configs[winner_config_index]); }
+
+  configs[2].state = config_state::Inactive;
```
Comment (lalo) [inline vowpalwabbit/core/src/reductions/details/automl/automl_oracle.cc:47]: reference valid_config_size, clear all except first two

File: vowpalwabbit/core/src/reductions/details/automl/automl_iomodel.cc [hunk 4]
```diff
@@ -34,21 +34,30 @@ void interaction_config_manager<config_oracle_impl, estimator_impl>::persist(met
...
-        metrics, "_amls_" + std::to_string(live_slot), verbose, _config_oracle._interaction_type);
-    estimators[live_slot].second.persist(metrics, "_sc_" + std::to_string(live_slot));
+        self_metrics, "_amls_" + std::to_string(live_slot), verbose, _config_oracle._interaction_type);
+    estimators[live_slot].second.persist(respective_champ_metrics, "_sc_" + std::to_string(live_slot));
     if (verbose)
     {
       auto& elements = _config_oracle.configs[estimators[live_slot].first.config_index].elements;
-      metrics.set_string("exclusionc_" + std::to_string(live_slot), VW::reductions::util::elements_to_string(elements));
+      nested_metrics.set_string(
+          "exclusionc_" + std::to_string(live_slot), VW::reductions::util::elements_to_string(elements));
```
Comment (lalo) [inline vowpalwabbit/core/src/reductions/details/automl/automl_iomodel.cc:49]: check config type

**Review by lalo — APPROVED** [review]

**Pull request merged**